### ISSUE TITLE
How to use Rematch Type definitions for Effects that return values?

### ISSUE THREAD

**Issue by danawoodman** [rematch/rematch#609]
Right now I am getting type exceptions with effects that return values, eg:

```ts
// model "someModel.ts"
export default createModel<SomeModelState>({
  state: { foo: true },
  effects: dispatch => ({
    async stuff() {
      const something = await someAsyncLib()
      return { something }
    },
    async other() {d
      // Fails here, does not use return type from `something` effect
      const { something } = await dispatch.someModel.stuff()
    }
}
```

Getting:

```
Type error: Property 'something' does not exist on type 'Action<any, any>'
```

I expect it to know that I am returning a value in the `something` method but it does not. 

How can I make sure the typing is detected correctly?

Thanks!

**Comment by semoal** [comment]
1º -
 I'm not sure if Rematch is ready to be used 100% in TypeScript.
2º -
You're doing an await to dispatch.someModel.stuff(), but you didn't defined any reducer for handle async action stuff.

Example:
``````ts
export default createModel<SomeModelState>({
  state: { foo: true, something: [] },
  reducers: {
    stuffReducer: (state: SomeModelState, payload: any): SomeModelState => {
      return {
        ...state,
        something: payload,
      }
    },

  },
  effects: {
    async stuff() {
      const something = await someAsyncLib()
      this.stuffReducer(something)
    },
    async other() {
      // now you have on state somthing, so you can do anything you want here.
    }
  }
}
```

**Comment by danawoodman** [comment]
Thanks @semoal.

1. That would be sad if it were the case 😢 

2. Reducers only handle change of state (which I am doing, my code snippet wasn't complete). It seems that Rematch doesn't know that I am returning a value from the effect which is the core issue I am addressing. Rematch supports returning from an effect which you can `await` on in your React component or wherever you want. I want it to know what I am returning in my effect so I don't get Typescript errors *and* I can use type checking on the effect in my components. 

I am using the Rematch Immer plugin FYI, not sure if that is relevant

**Comment by canyavall** [comment]
@danawoodman 
I checked it in my code and I didn't have any issue, actually
´´´
    async fetchChannels() {

        this.setChannelsLoading(true)
        try {
            const url = "tv/channels"
            const data = await getRequestIB<ChannelsCenterChannel>(url, headers)
            return data
        } catch (e) {
            Logger.error("fetchChannels: " + e.message)
            dispatch.channelscenter.setChannelsRequestError(e)
            dispatch.channelscenter.setChannelsLoading(false)
        }

    },
´´´
 try {
            const data = await dispatch.channelscenter.fetchChannels()
        } catch (e) {
´´´

´´´

**Comment by canyavall** [comment]
@danawoodman I just faced the same issue, for some reason now the code complains...

I think there are quite a lot of issues with types and meanwhile they don't fix it (or someone that erally knows about it) I had to use :  //@ts-ignore

**Comment by bagr001** [comment]
@ShMcK This used to work in previous version. Now, the return value of the promise is always undefined. This is a cool "feature" I really love on rematch.

**Comment by semoal** [comment]
I'll check it guys as soon as possible.

**Comment by derrickbeining** [comment]
@bagr001 , I just notice that too. I know I just recently used Rematch where `effects` return the `Promise` for the value they `return`ed. But now it doesn't. Looks like there was a release on npm about a month ago. I tried installing a few previous version too though, and they all seem to return `Promise<undefined>` as well. Does anyone know the last version that allowed the caller to consume the return value of effects?

**Comment by bagr001** [comment]
@derrickbeining I am using `@rematch/core@1.0.6` which seem to still have this feature.

**Comment by derrickbeining** [comment]
@bagr001 I tried that one and I'm still getting the same `Promise<undefined>` behaviour. Ugh. What is happening.

**Comment by derrickbeining** [comment]
@bagr001 , eureka! I've discovered that (at least for me) the issue was coming from `@rematch/loading`, not `@rematch/core`. I was using the loading pluging (as usual) and when I commented it out, my effects started returning `Promise<expected_value>`. I suggest you try that as well.

**Comment by derrickbeining** [comment]
@bagr001 okay, so apparently `@rematch/loading` had a bogus release a month ago that reverted `latest` from `1.1.3` to `0.5.1`, so that's why the loading plugin is breaking stuff. Guess I'll open a separate issue for that.

**Comment by derrickbeining** [comment]
https://github.com/rematch/rematch/issues/629#issue-429764261

**Comment by danawoodman** [comment]
This still breaks for me using the latest versions of everything with: 

```
TS2345: Argument of type '(dispatch: RematchDispatch<Models>) => { toggleCollapse: RematchDispatcher<void, void> | (((action: Action<any, any>) => Dispatch<Action<any, any>>) & ((action: Action<any, void>) => Dispatch<...>) & ((...args: any[]) => Action<...>)) | (((action: Action<...>) => Dispatch<...>) & ... 1 more ... & ((payload: any) => ...' is not assignable to parameter of type 'MapDispatchToPropsParam<{ toggleCollapse: RematchDispatcher<void, void> | (((action: Action<any, any>) => Dispatch<Action<any, any>>) & ((action: Action<any, void>) => Dispatch<Action<any, void>>) & ((...args: any[]) => Action<...>)) | (((action: Action<...>) => Dispatch<...>) & ... 1 more ... & ((payload: any) => A...'.
  Type '(dispatch: RematchDispatch<Models>) => { toggleCollapse: RematchDispatcher<void, void> | (((action: Action<any, any>) => Dispatch<Action<any, any>>) & ((action: Action<any, void>) => Dispatch<...>) & ((...args: any[]) => Action<...>)) | (((action: Action<...>) => Dispatch<...>) & ... 1 more ... & ((payload: any) => ...' is not assignable to type 'MapDispatchToPropsFactory<{ toggleCollapse: RematchDispatcher<void, void> | (((action: Action<any, any>) => Dispatch<Action<any, any>>) & ((action: Action<any, void>) => Dispatch<Action<any, void>>) & ((...args: any[]) => Action<...>)) | (((action: Action<...>) => Dispatch<...>) & ... 1 more ... & ((payload: any) =>...'.
    Types of parameters 'dispatch' and 'dispatch' are incompatible.
      Type 'Dispatch<Action<any>>' is not assignable to type 'RematchDispatch<Models>'.
        Type 'Dispatch<Action<any>>' is not assignable to type 'ExtractRematchDispatchersFromModels<Models> & ((action: Action<void, void>) => Dispatch<Action<void, void>>) & ((action: Action<void, void>) => Dispatch<Action<void, void>>) & ((...args: any[]) => Action<...>) & Dispatch<...>'.
          Type 'Dispatch<Action<any>>' is not assignable to type 'ExtractRematchDispatchersFromModels<Models>'.
            Index signature is missing in type 'Dispatch<Action<any>>'.
```

This is using code exactly like in the [Typescript recipe](https://github.com/rematch/rematch/blob/master/docs/recipes/typescript.md).

@semoal are you able to get the example working without error?

**Comment by Mkaranter** [comment]
Having the same issue in my project. Strictly followed ts + rematch and still getting the same error.

```Argument of type '({ view, form }: RematchDispatch<typeof models>) => { toggleUserForm: ((action: Action<any, void>) => Dispatch<Action<any, void>>) & ((action: Action<...>) => Dispatch<...>) & ((...args: any[]) => Action<...>); addQuestion: ((action: Action<...>) => Dis...' is not assignable to parameter of type 'MapDispatchToPropsParam<{ toggleUserForm: ((action: Action<any, void>) => Dispatch<Action<any, void>>) & ((action: Action<any, void>) => Dispatch<Action<any, void>>) & ((...args: any[]) => Action<any, any>); addQuestion: ((action: Action<...>) => Dispatch<...>) & ... 1 more ... & ((...args: any[]) => Action<...>); g...'.
  Type '({ view, form }: RematchDispatch<typeof typeof models>) => { toggleUserForm: ((action: Action<any, void>) => Dispatch<Action<any, void>>) & ((action: Action<...>) => Dispatch<...>) & ((...args: any[]) => Action<...>); addQuestion: ((action: Action<...>) => Dis...' is not assignable to type 'MapDispatchToPropsFactory<{ toggleUserForm: ((action: Action<any, void>) => Dispatch<Action<any, void>>) & ((action: Action<any, void>) => Dispatch<Action<any, void>>) & ((...args: any[]) => Action<any, any>); addQuestion: ((action: Action<...>) => Dispatch<...>) & ... 1 more ... & ((...args: any[]) => Action<...>);...'.
    Types of parameters '__0' and 'dispatch' are incompatible.
      Type 'Dispatch<Action<any>>' is not assignable to type 'RematchDispatch<typeof models>'.
        Type 'Dispatch<Action<any>>' is not assignable to type 'ExtractRematchDispatchersFromModels<typeof models> & ((action: Action<void, void>) => Dispatch<Action<void, void>>) & ((action: Action<...>) => Dispatch<...>) & ((...args: any[]) => Action<...>) & Dispatch<...>'.
```

@danawoodman looks like this is the same issue. Did you find reasonable solution? 

<img width="328" alt="Zrzut ekranu 2019-07-25 o 21 39 03" src="https://user-images.githubusercontent.com/20187678/61903426-c0cf3800-af24-11e9-8376-ab53d4a17344.png">

**Comment by TomasTarasovic** [comment]
I had the same issue as @Mkaranter and @danawoodman.
I am using "@rematch/core": "^1.4.0", "react-redux": "^7.2.0", "typescript": "~3.7.2"
Is there any working solution without errors?
I would like to use connect not hooks.
@semoal 
Thanks, guys!

**Comment by barakoles** [comment]
dont let it staaapaale

**Comment by semoal** [comment]
> dont let it staaapaale

This is fixed for a long time ago, since we moved to v2.0 version
